### PR TITLE
Generate candidate sets for all users at once.

### PR DESCRIPTION
**Pull request by vansika**
# Problem
We have been generating candidate sets for users one at a time which leads to spark errors. 

# Solution
Generate candidate sets for all users at once.

### CODE REVIEW

**Comment by vansika** [comment]
If this is still a mess, let me know :)

File: listenbrainz_spark/recommendations/candidate_sets.py [hunk 1]
```diff
@@ -24,211 +24,217 @@
...
-        artists[row.top_artist_name].append(row.similar_artist_name)
-    return artists
+    user_data = defaultdict(list)
+    for row in top_similar_artists_df.collect():
+        user_data[row.user_name].append((row.top_artist_name, row.similar_artist_name, row.score))
+    return user_data
+
 
-def save_candidate_html(user_data):
+def save_candidate_html(user_data, ti):
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/candidate_sets.py:222]: ti is not a good name for this arg. time_to_generate ?

Comment (vansika) [reply listenbrainz_spark/recommendations/candidate_sets.py:222]: Ti is not the time to generate, it is the initial time, we subtract it from final time and get the interval. 

Maybe time_initial?

Comment (mayhem) [reply listenbrainz_spark/recommendations/candidate_sets.py:222]: sure

File: listenbrainz_spark/recommendations/candidate_sets.py [hunk 2]
```diff
@@ -24,211 +24,217 @@
...
-def get_top_artists(df, user_name):
-    """ Get top artists listened to by the user.
+def get_top_artists(mapped_listens_subset):
+    """ Get top artists listened to by users who have a listening history in
+        the past X days where X = RECOMMENDATION_GENERATION_WINDOW.
 
         Args:
-            df (dataframe): Dataframe to containing user history.
-            user_name (str): Name of the user.
+            df (dataframe): A subset of mapped_df containing user history.
```
Comment (shivam-kapila) [inline listenbrainz_spark/recommendations/candidate_sets.py:72]:
```suggestion
            mapped_listens_subset (dataframe): A subset of mapped_df containing user history.
```

Comment (vansika) [reply listenbrainz_spark/recommendations/candidate_sets.py:72]: Thanks. I have already made these changes in a separate commit. Will puch it soon :)

**Pull request merged**